### ISSUE TITLE
larx/stcx aware refcounting

### ISSUE THREAD

**Issue by npiggin** [linuxppc/issues#421]
refcount_inc/add_not_zero is an important primitive used in performance critical code.

This is implemented with generic atomics using read / cmpxchg pattern, which is suboptimal for powerpc. This should be optimised. Using an existing atomic primitive like atomic_add_unless would be ideal, to avoid having to add a new arch specific primitive or override.

**Comment by chleroy** [comment]
FWIW see issue #87 , possibility to implement arch specific version of refcount was removed in commit https://github.com/linuxppc/linux/commit/fb041bb7c0a918b95c6889fc965cdc4a75b4c0ca

**Comment by npiggin** [comment]
Yeah, we could add something back though if necessary. Or even better we might be able to just use existing atomic_add_unless in the core code I think.

**Comment by mpe** [comment]
There was a lot of noise around the original refcount work, and the performance impact it would have. AFAIR in the end no one could show any meaningful performance impact and that's why the optimised version was removed. Which is not to say we can't optimise it, but some folks will be inclined to push back based on the previous experience.

**Comment by npiggin** [comment]
It's hard because microbenchmarks won't show something like this because the cacheline will always be exclusive. On a big system and a real workload where we're running out of coherency bandwidth, then minimizing coherency operations could help.
